pipeline/transform.py

```diff
@@ -0,0 +1,131 @@
+    """
+    try:
+        return reading['origin_location'][index]
+    except KeyError:
+        return None
+
+
+def get_scientific_name(reading: dict) -> str | None:
+    """
+    Fetches scientific name of the plant
+    """
+    try:
+        return reading['scientific_name'][0]
+    except KeyError:
+        return None
+
+
+def get_origin_region(reading: dict) -> str | None:
+    """
+    Fetches the region of where the plant originated from
+    """
+    try:
+        return reading['origin_location'][4].split('/')[0]
+    except KeyError:
+        return None
+
+
+def get_details(reading: dict, detail: str) -> str | int | None:
+    """
+    Fetches specific details by searching through the keys
+    """
+    try:
+        return reading[detail]
+    except KeyError:
+        return None
+
+
+def botanist_details(reading: list[dict]) -> dict:
+    """
+    Returns a dataframe of botanist details
+    """
+    return {'name': get_botanist_detail(reading, 'name'),
+            'email': get_botanist_detail(reading, 'email'),
+            'phone_no': get_botanist_detail(reading, 'phone')
+            }
+
+
+def plant_details(reading: list[dict]) -> dict:
+    """
+    Returns a dataframe of plant details
+    """
+    return {'plant_id': get_details(reading, 'plant_id'),
+                    'plant_name': get_details(reading, 'name'),
+                    'scientific_name': get_scientific_name(reading),
+                    'origin_longitude': get_origin_detail(reading, 1),
+                    'origin_latitude': get_origin_detail(reading, 0),
+                    'origin_town': get_origin_detail(reading, 2),
+                    'origin_country_code': get_origin_detail(reading, 3),
+                    'origin_region': get_origin_region(reading)}
+
+
+def plant_readings(reading: list[dict]) -> dict:
+    """
+    Returns a dataframe of reading details
+    """
+    return {
+            'soil_moisture': get_details(reading, 'soil_moisture'),
+            'temperature': get_details(reading, 'temperature'),
+            'last_watered': get_details(reading, 'last_watered'),
+            'recording_taken': get_details(reading, 'recording_taken'),
+            'name': get_botanist_detail(reading, 'name'),
+            'plant_name': get_details(reading, 'name'),
+        }
+
+def group_data(readings: list[dict]) -> tuple[list, list, list]:
+    """
+    Iterates through the readings and organises different parameters
+    into the relevant list
+    """
+    plant = []
+    botanist = []
+    plant_reading = []
+
+    for reading in readings:
+        plant.append(plant_details(reading))
+        botanist.append(botanist_details(reading))
+        plant_reading.append(plant_readings(reading))
+
+    return plant, botanist, plant_reading
+
+
+def convert_to_dataframe(readings: list[dict]) -> pd.DataFrame:
+    """
+    Converts a list into a dataframe
+    """
+    return pd.DataFrame(readings)
```

Comment:
What if readings isn't a list of dicts? Consider adding some type checking at the start of the function. E.g.
```suggestion
def convert_to_dataframe(readings: list[dict]) -> pd.DataFrame:
    """
    Converts a list into a dataframe
    """
    if not isinstance(readings, list):
        raise TypeError('readings must be of type list')
    if not all(isinstance(value, dict) for value in readings):
        raise TypeError('entries into a DataFrame must be of type dict')
    return pd.DataFrame(readings)
```